{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Umair Khan\n",
    "\n",
    "Weighted Gene Co-expression Network Analysis of CD8 and CD4 T Cells of Infected Covid-19 Patients\n",
    "\n",
    "\n",
    "Copy of R Markdown to read in RDS File containing metadata and get the scaled gene expression data matrix of highly variable genes.\n",
    "\n",
    "Here I use the Seurat package to subset the data to remove unwanted genes and cells, extract different T-Cells in the expression matrix, normalize, find variable genes, and scale. \n",
    "\n",
    "memory.limit(30000)\n",
    "library(Seurat)\n",
    "pbmcs <- readRDS(\"C:\\\\Users\\\\zulfi\\\\Desktop\\\\blish_covid.seu.rds\")\n",
    "DefaultAssay(pbmcs) <- \"RNA\"\n",
    "\n",
    "pbmcs <- subset(pbmcs, subset = nFeature_RNA < 4000 & nFeature_RNA > 500 & percent.mt < 10 & percent.mt > 2.5)\n",
    "\n",
    "\n",
    "Idents(object = pbmcs) <- \"cell.type\"\n",
    "TCells <- subset(pbmcs,idents = c(\"CD4m T\",\"CD4n T\",\"CD8m T\",\"CD8eff T\"))\n",
    "\n",
    "\n",
    "covTcells <- NormalizeData(covTcells,normalization.method = \"LogNormalize\",scale.factor = 10000, assay = \"RNA\")\n",
    "\n",
    "covTcells <- FindVariableFeatures(covTcells, selection.method = \"vst\", nfeatures = 2000,assay = \"RNA\")\n",
    "\n",
    "covTcells <- ScaleData(covTcells,assay = \"RNA\",vars.to.regress = \"percent.mt\",do.scale = TRUE,\n",
    "                       do.center = TRUE)\n",
    "\n",
    "library(data.table)\n",
    "covMat <- t(covTcells@assays$RNA@scale.data)\n",
    "covMat <- as.data.frame(covMat)\n",
    "df <- fwrite(covMat,file=\"covTcells.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats.mstats import rankdata \n",
    "import community as community_louvain\n",
    "\n",
    "class CoexpressionNetwork:\n",
    "    \n",
    "    '''\n",
    "    Python version of R's WGCNA using a biweight midcorrelation for the similarity matrix and topological overlap \n",
    "    measure for the network proximity measure. I convert the topological overlap measure dissimilarity matrix to a \n",
    "    networkx object and implement the Louvain method for finding network modularity. Last, I find hub genes for each \n",
    "    cluster membership found by Louvain. \n",
    "    \n",
    "    Metadata of single-cell peripheral blood mononuclear cells of infected Covid Patients from the Human\n",
    "    Covid-19 Cell Atlas.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    def __init__(self,dgeMatrix,softThres,transform):\n",
    "        \n",
    "        '''\n",
    "        Class Parameters:\n",
    "        \n",
    "        dgeMatrix: CSV File containing a scaled gene expression of highly variable genes only.\n",
    "        softThres: Power to take for soft thresholding \n",
    "        transform: Number to transform matrix so the signed matrix can be on the 0-1 scale.\n",
    "        \n",
    "        '''\n",
    "        \n",
    "        self.dgeMatrix = pd.read_csv(dgeMatrix)\n",
    "        self.softThres = softThres\n",
    "        self.transform = transform\n",
    "        self.mappings = { index:column for index,column in enumerate(self.dgeMatrix) }\n",
    "        self.dgeMatrix = self.dgeMatrix.to_numpy(dtype='float')\n",
    "        self.bicorMat = np.zeros([self.dgeMatrix.shape[1],self.dgeMatrix.shape[1]])\n",
    "        self.adjMat = np.array([self.dgeMatrix.shape[1],self.dgeMatrix.shape[1]])\n",
    "        self.tomMat = np.zeros([self.dgeMatrix.shape[1],self.dgeMatrix.shape[1]])\n",
    "        self.corNet = nx.Graph()\n",
    "        self.partitions = {}\n",
    "        \n",
    "    def bicorMatrix(self):\n",
    "        \n",
    "        '''\n",
    "        Computes the biweight midcorrelation matrix.\n",
    "        Algorithm from : https://en.wikipedia.org/wiki/Biweight_midcorrelation\n",
    "        \n",
    "        '''\n",
    "        X = self.dgeMatrix - np.median(self.dgeMatrix,axis=0)\n",
    "        \n",
    "        U = X / (9*np.median(np.abs(X),axis=0))\n",
    "        \n",
    "        W = ((1 - U ** 2) ** 2) * ((1-np.abs(U)) > 0)\n",
    "    \n",
    "        E = X*W\n",
    "                \n",
    "        for i in range(self.dgeMatrix.shape[1]):\n",
    "            \n",
    "            for j in range(self.dgeMatrix.shape[1]):\n",
    "                \n",
    "                self.bicorMat[i,j] = np.dot(E[:,i],E[:,j]) / (np.sqrt(np.sum(E[:,i] ** 2)) * np.sqrt(np.sum(E[:,j] ** 2)))\n",
    "                  \n",
    "        return self.bicorMat\n",
    "                              \n",
    "\n",
    "    def adjacencyMatrix(self):\n",
    "        \n",
    "        '''\n",
    "        Creates a weighted adjacency matrix/network by transforming the biweight midcorrelation matrix\n",
    "        to a signed weighted co-expression graph on the 0-1 scale and then applying a power function \n",
    "        for soft thresholding, as opposed to hard thresholding for the adjacency matrix. For transforming,\n",
    "        multiply correlation matrix by 0.5 then add by 0.5. Then we take the power of the result by a softhresholding\n",
    "        value.\n",
    "        \n",
    "        '''\n",
    "        \n",
    "        self.adjMat = np.power(self.bicorMat*self.transform + self.transform, self.softThres)\n",
    "            \n",
    "        np.fill_diagonal(self.adjMat,0)\n",
    "        \n",
    "        return self.adjMat\n",
    "    \n",
    "   \n",
    "    def tomMatrix(self):\n",
    "        \n",
    "        '''\n",
    "        Computes the topological overlap measurement(TOM) to make a TOM dissimilarity matrix.\n",
    "        Algorithm from : https://dibernardo.tigem.it/files/papers/2008/zhangbin-statappsgeneticsmolbio.pdf\n",
    "        \n",
    "        '''\n",
    "        \n",
    "        L = np.matmul(self.adjMat,self.adjMat)\n",
    "        \n",
    "        K = self.adjMat.sum(axis=1)\n",
    "                        \n",
    "        for i in range(self.adjMat.shape[0]):\n",
    "            \n",
    "            for j in range(i+1,self.adjMat.shape[0]):\n",
    "                \n",
    "                numerator = L[i,j] + self.adjMat[i,j]\n",
    "                \n",
    "                denominator = min(K[i],K[j]) + 1 - self.adjMat[i,j]\n",
    "                \n",
    "                self.tomMat[i,j] = numerator/denominator\n",
    "                                \n",
    "        self.tomMat += self.tomMat.T\n",
    "          \n",
    "        np.fill_diagonal(self.tomMat,1)\n",
    "        \n",
    "        self.tomMat = 1 - self.tomMat\n",
    "            \n",
    "        return self.tomMat\n",
    "        \n",
    "            \n",
    "    def corrNetwork(self):\n",
    "        \n",
    "        '''\n",
    "        Convert weighted co-expression adjacency network to a networkx graph object and relabel nodes to corresponding \n",
    "        integer:gene mapping.\n",
    "        \n",
    "        '''\n",
    "\n",
    "        self.corNet = nx.from_numpy_array(self.tomMat, parallel_edges = False)\n",
    "        \n",
    "        self.corNet = nx.relabel.relabel_nodes(self.corNet,self.mappings)\n",
    "                \n",
    "        return self.corNet\n",
    "    \n",
    "    def louvain(self):\n",
    "        \n",
    "        '''\n",
    "        Use the Louvain method for finding network modules. From python-louvain and networkx. \n",
    "        \n",
    "        '''\n",
    "        \n",
    "        self.partitions = community_louvain.best_partition(self.corNet,weight='weight',resolution=1.0)\n",
    "        \n",
    "        return self.partitions\n",
    "\n",
    "\n",
    "    def hubGenes(self):\n",
    "        \n",
    "        '''\n",
    "        Find hub genes within each cluster. geneClusters is a dictionary where a list of genes are keyed their cluster id \n",
    "        membership. networkClusters is a list containing networkx graph objects only containing genes by corresponding\n",
    "        cluster id. I then use nx.hits_numpy to find hub genes for each graph objects and return the top 20 hub genes.\n",
    "        \n",
    "        Hubs found using HITS algorithm using networkx's built in function\n",
    "    \n",
    "        '''\n",
    "        \n",
    "        geneClusters,networkClusters = {},[]\n",
    "        \n",
    "        for genes,cluster in self.partitions.items():\n",
    "            \n",
    "            if cluster not in geneClusters:\n",
    "                \n",
    "                geneClusters[cluster] = [] \n",
    "                \n",
    "            geneClusters[cluster].append(genes)\n",
    "                        \n",
    "        for index in geneClusters:\n",
    "\n",
    "            networkClusters.append(nx.Graph(((u,v,w) for u,v,w in self.corNet.edges(data=True) if u in geneClusters[index])))\n",
    "        \n",
    "        hubs = [ nx.hits_numpy(n) for n in networkClusters ]\n",
    "        \n",
    "        output =  [ sorted(j[0].items(),key=lambda x:x[1],reverse = True)[:20] for j in hubs ]\n",
    "        \n",
    "        return output\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[('CANX', 0.001684423549335662),\n",
       "  ('FTL', 0.0016217710091694219),\n",
       "  ('ARPC5', 0.0015234710387957446),\n",
       "  ('MALAT1', 0.0014907247540372804),\n",
       "  ('ANXA1', 0.0014887388922999738),\n",
       "  ('CCL5', 0.0014694832332506273),\n",
       "  ('CALR', 0.0013988704984717977),\n",
       "  ('DEK', 0.001354717558426897),\n",
       "  ('CCT3', 0.0013458539423112858),\n",
       "  ('ITGA4', 0.0013430360986003789),\n",
       "  ('ATP5A1', 0.001334474656488923),\n",
       "  ('CD8A', 0.0013304131752991959),\n",
       "  ('GPRIN3', 0.0013296992125983666),\n",
       "  ('ANP32E', 0.0013171288618258998),\n",
       "  ('EFHD2', 0.0013106884630893244),\n",
       "  ('BIRC3', 0.0013102212436055605),\n",
       "  ('CTSS', 0.001309039084519524),\n",
       "  ('HMGB1', 0.0013082055981069555),\n",
       "  ('DNAJA1', 0.0013045927815311768),\n",
       "  ('ITGB1', 0.001301820813357916)],\n",
       " [('ACTB', 0.0009756952815166152),\n",
       "  ('ACTR3', 0.0009730277682695686),\n",
       "  ('ACTG1', 0.0009692350881815931),\n",
       "  ('ARPC2', 0.0009625481428736085),\n",
       "  ('CAP1', 0.0009615419476105815),\n",
       "  ('ENO1', 0.0009574758452606727),\n",
       "  ('HNRNPA2B1', 0.0009533975848766701),\n",
       "  ('HSP90AA1', 0.0009525767610006365),\n",
       "  ('HSP90AB1', 0.000947963448508348),\n",
       "  ('HSPA5', 0.0009432312972583765),\n",
       "  ('IL7R', 0.0009417365124926998),\n",
       "  ('HSPA8', 0.000940852685609516),\n",
       "  ('LCP1', 0.0009407079270440245),\n",
       "  ('HSP90B1', 0.0009397320566495209),\n",
       "  ('LTB', 0.0009357189818681486),\n",
       "  ('MT-CO2', 0.0009303705475395712),\n",
       "  ('MT-CO3', 0.0009303548758158427),\n",
       "  ('MT-CYB', 0.0009298737536982775),\n",
       "  ('MT-ATP6', 0.0009279132947864521),\n",
       "  ('MT-CO1', 0.0009278470433207501)]]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cn = CoexpressionNetwork(dgeMatrix='covTcells.csv',softThres=12,transform=0.5)\n",
    "cn.bicorMatrix()\n",
    "cn.adjacencyMatrix()\n",
    "cn.tomMatrix()\n",
    "cn.corrNetwork()\n",
    "cn.louvain()\n",
    "cn.hubGenes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "Louvain modularity identifies two clusters taking the topological overlap measure matrix as input. Gene functions taken from genecards.org, unless otherwise cited.\n",
    "\n",
    "Module 1 identifies genes associated in neuropathogeneis including GPRIN3(implicated in neurite growth), ANP32E(implicated in cerebellar genesis and synaptogenesis), and FTL induced neurodegeneration due to oxidative stress in the ETC. Module 1's genes with Metastatic properties include MALAT1(a well known long spliced non-coding RNA that been noted to have a role in transcriptional regulation of genes involved in cancer metastasis, cell migration, and upregulation of cancerous tissues by proliferation, CALR which produces the calreticulin protein involved in cell migration,cell adhesion, gene activity,and cell growth(1) through calcium regulation, and BIRC3 modulating inflammatory signaling/immunity, mitogenic kinase signaling,cell proliferation, cell migration, and metastasis. ITGA4 and its interaction with ITGAB1 has been studied to influence cell migration and motility and are receptors for fibronectin known to function for cell differentiation, migration and wound healing. Integrin family proteins also exhibit properties to induce cell mobility residing near the extracellular matrix, and this along with CTSS that is actively involved in blood vessel angiogenesis(2) suggest the presence of these family of proteins at the site of blood vessel differentiation. \n",
    "\n",
    "Module 2 identifies genes relating to actin polymerization including ACTB, ACTR3, ACTG1,ARPC2,CAP1(implicated in complex developent and morphological processes). ENO1 has suggested functions  in the intravascular and pericellular fibrinolytic system due to its ability to serve as a receptor and activator of plasminogen on the cell surface of several cell-types such as leukocytes and neurons. Isoforms of LCP1 are known to have replication potential in epithelial cells, fibroblasts, and endothelial cells and LTB has been involved in cancer metastasis. HSP90AA1, HSP90AB1, HSP90B1,HSPA5, and HNRNPA2BI(pre-mRNA processing) are mostly molecular chaperone proteins with functions ranging from protein quality control, folding, and morphological evolution, structure, maintenence, where HSP90 co modulate transcription in three levels including physiological cues, steady-state levels in epigenetic modifiers, and gene expression through the eviction of histones on promoter regions. MT-CO1,MT-CO2,MT-CO3, MT-CYB, and MT-ATP6,  constitute mitochondrial genes that have functions including regulation of the electron transport chain, oxidative phosphorylation, and aerobic metabolism. Defencies in MT-ATP6 and MT-CYB are both known to have an effect on neurodegeneration and cardiovascular health, and these set of genes implicate an inflammation that leads to mitochondrial oxidative damage by Sars-Cov-2(3).\n",
    "\n",
    "Key Takeaway: Hub genes found from gene expression data of T cells from infected Covid-19 patients suggest a multitude of problems that include: \n",
    "\n",
    "1. actin polymerization\n",
    "2. cell migration and metastasis properties\n",
    "3. mitochondrial level cardiovascular damage\n",
    "4. implications of neuron differentiation and growth\n",
    "5. damage to mitochondrial cytochrome subunits\n",
    "\n",
    "Hub genes mentioned and listed maybe of clinical use to investigate for drug therapeutics to combat Sars-Cov-2 in terms observing pathways in actin growth to reduce virus protrusion and drugs working to improve mitochondrial function and reduce oxidative stress and mitochondrial disfunction.  \n",
    "\n",
    "\n",
    "Sources:\n",
    "\n",
    "1. https://ghr.nlm.nih.gov/gene/CALR\n",
    "2. https://cancerres.aacrjournals.org/content/69/10/4537\n",
    "3. https://www.sciencedirect.com/science/article/pii/S1567724920301380"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
