{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pareto Archetypal Analysis of PBMCs among Infected Covid-19  Patients\n",
    "\n",
    "Sars-Cov-2 not only has been shown to damage the respiratory tract and lung tissues, but it has a substantial effect on neurological pathologies, the cardiovascular system and thrombotic effects throughout different organs particularly organs with high abundance of ACE2 receptors. This strongly suggests the virus can move past the lungs and has a role within the endothelial cell linings of the blood vessels. Mechanisms behind the virus's invasive properties and ability to cause blood vessel and brain damage is explored to determine proteins that can be considered for drug targets to mitigate virus spread.\n",
    "\n",
    "Archetypal analysis is an unsupervised learning method to identify archetypes or extreme points in multidimensional data. The archetypes are put on a Pareto front, that is, no state can be made better without making one worse(Pareto optimality) where these arcs are contained within an efficient representation of a polytope using the principal convex hull algorithm. In a biological context, archetypes found in gene expression data may indicate highly expressed genes, that is, certain tradeoffs are made within cell and genes to achieve an optimal biological function determined by the archetype.\n",
    "\n",
    "Here, I implement a Pareto Archetypal Analysis using R's ParetoTI package to find key biological processes or archetypes within peripheral blood mononuclear cells(PBMC) of infected Covid-19 Patients. I use Python's gseapy package to perform a gene set enrichment analysis at each archetype, and observe protein-protein interactions(PPI) at each archetype using STRING. Another Archetypal Analysis is done on CD4 T cells determined to be the cell type closest to the specific archetype of interest along with a GSEA and PPI network analysis.\n",
    "\n",
    "Single-cell metadata of PBMCs taken from the Human Covid-19 Cell Atlas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following steps in R include:\n",
    "\n",
    "1. Reading in the RDS File containing PBMCs metadata\n",
    "\n",
    "2. Subsetting/Filtering cells and genes by visualizing violin and correlation plots. Violin plots for mitochondrial contamination, gene counts, and molecule counts are shown to observe where most of the data lies. Correlating molecule count with mitochondria contamination and gene counts was done to subset my data; I include genes that have a strong correlation with the molecular count. Although I include quite a significant amount of mitochondrial contamination, most of the data lies between 2.5 and 10 percent contamination; a more strict threshold will not give me biological signal. However, I will regress this out when I scale my data later on.\n",
    "\n",
    "3. Normalization, finding variable features, and scaling using Seurat functions. Percent of mitochondria is regressed out.\n",
    "\n",
    "4. Principal Component Analysis and determining how many principal components to use based on a skree plot. 10 principal components were used as they contribute to the most variation.\n",
    "\n",
    "5. Finding the optimal number of archetypes to use based on t-ratio and variation plots. Although I find that 8 archetypes contribute to most of the variation in my dataset, the 6th archetype has the highest t-ratio, that is the most signficant, therefore, 6 archetypes were used.\n",
    "\n",
    "6. Run principal convex hull algorithm with 6 archetypes. \n",
    "\n",
    "7. I used other functions within ParetoTI to find enriched genes at each archetype using a Wilcoxon one versus all test where we use a Euclidean distance matrix for the features and find a decreasing function from each feature to the archetype. I adjust p values for multiple comparisons using the Benjamin-Hochberg procedure.\n",
    "\n",
    "8. The enriched genes were written to a data table to be used for a gene set enrichment analysis to discover biological processes using Pythonâ€™s gseapy package, in which a pre ranked GSEA was run where the test statistic for gene ranks was the Wilcoxon mean log ratio. This automatically puts gene rankeed by archetype distance in order."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# R code for the Pareto Archetype Analysis \n",
    "\n",
    "\n",
    "library(ParetoTI)\n",
    "\n",
    "library(data.table)\n",
    "\n",
    "pbmcs <- readRDS(\"C:\\\\Users\\\\zulfi\\\\Desktop\\\\blish_covid.seu.rds\")\n",
    "\n",
    "library(Seurat)\n",
    "\n",
    "DefaultAssay(pbmcs) <- \"RNA\"\n",
    "\n",
    "VlnPlot(pbmcs,features = c(\"nFeature_RNA\",\"nCount_RNA\",\"percent.mt\"),ncol = 3)\n",
    "\n",
    "\n",
    "plot1 <- FeatureScatter(pbmcs,feature1 = \"nCount_RNA\",feature2 = \"percent.mt\")\n",
    "\n",
    "plot2 <- FeatureScatter(pbmcs, feature1 = \"nCount_RNA\", feature2 = \"nFeature_ RNA\")\n",
    "\n",
    "plot1 + plot2\n",
    "\n",
    "<img src=\"cmp.png\" width=\"800\" />\n",
    "\n",
    "pbmcs <- subset(pbmcs, subset = nFeature_RNA < 4000 & nFeature_RNA > 500 & percent.mt < 10 & percent.mt > 2.5)\n",
    "\n",
    "plot3 <- FeatureScatter(pbmcs, feature1 = \"nCount_RNA\", feature2 = \"nFeature_ RNA\")\n",
    "\n",
    "print(plot3)\n",
    "\n",
    "<img src=\"cmp2.png\" width=\"800\" />\n",
    "\n",
    "covid <- subset(pbmcs,Status==\"COVID\")\n",
    "\n",
    "memory.limit(30000)\n",
    "\n",
    "covid <- NormalizeData(covid,normalization.method = \"LogNormalize\",scale.factor = 10000, assay = \"RNA\")\n",
    "\n",
    "covid <- FindVariableFeatures(covid, selection.method = \"vst\", nfeatures = 2000,assay = \"RNA\")\n",
    "\n",
    "covid <- ScaleData(covid,assay = \"RNA\",vars.to.regress = \"percent.mt\")\n",
    "\n",
    "covid <- RunPCA(covid, features = VariableFeatures(object = covid),assay = \"RNA\",npcs = 50)\n",
    "\n",
    "eigs <- covid@reductions$pca@stdev**2\n",
    "\n",
    "props <- eigs/sum(eigs)\n",
    "\n",
    "plot(props,ylab = \"Proportion of Variation\",xlab=\"Principal Components\")\n",
    "\n",
    "<img src=\"PCA.png\" width=\"800\" />\n",
    "\n",
    "\n",
    "pca <- data.frame(covid@reductions$pca\\@cell.embeddings)\n",
    "\n",
    "pca <- pca[c(1:10)]\n",
    "\n",
    "pca <- data.matrix(pca)\n",
    "\n",
    "pca <- t(pca)\n",
    "\n",
    "pcha <- k_fit_pch(pca,ks = 3:8,maxiter = 500,seed=345)\n",
    "\n",
    "plot_arc_var(pcha, type = \"varexpl\", point_size = 2)\n",
    "\n",
    "plot_arc_var(pcha, type = \"t_ratio\", point_size = 2)\n",
    "\n",
    "<img src=\"var.png\" width=\"800\" />\n",
    "\n",
    "<img src=\"trat.png\" width=\"800\" />\n",
    "\n",
    "\n",
    "arc <- fit_pch(pca,noc= 6,conv_crit = 1e-06,maxiter = 500,volume_ratio= \"t_ratio\")\n",
    "\n",
    "plt_arc <- plot_arc(arc_data = arc, data = pca)\n",
    "\n",
    "print(plt_arc)\n",
    "\n",
    "\n",
    "<img src=\"pbmcsArcs.png\" width=\"800\" />\n",
    "\n",
    "\n",
    "mergeDat <- merge_arch_dist(arc_data = arc,data = pca,dist_metric = c(\"euclidean\", \"arch_weights\")[1],feature_data = covid@assays$RNA\\@scale.data)\n",
    "\n",
    "enrGenes <- find_decreasing_wilcox(mergeDat\\\\$data,mergeDat\\\\$arc_col,\n",
    "features= mergeDat$features_col,bin = 0.1,method = \"BioQC\")\n",
    "\n",
    "genesAtArcs <- get_top_decreasing(summary_genes = enrGenes,summary_sets = NULL, cutoff_genes = 0.001,cutoff_metric = \"wilcoxon_p_val\",p.adjust.method = \"BH\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform a pre-ranked Gene Set Enrichment Analysis at each Archetype "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "This short script takes in a data table in a csv file and performs a preranked GSEA. I use the top n genes closest to the\n",
    "archetype for the enrichment analysis and output the top 10 biological processes ranked by the normalized enrichment score;\n",
    "the overrepresented upregulated processes are shown. \n",
    "\n",
    "The GSEA uses pre ranked gene lists with the top n genes and the wilcoxon mean log ratio as the test statistic. \n",
    "\n",
    "Class Attributes:\n",
    "\n",
    "fileObj = Data table file of archetypes and corresponding enriched genes\n",
    "nGenes = Number of genes to use within each archetype for the gene set enrichment analysis\n",
    "\n",
    "'''\n",
    "\n",
    "class ParetoGSEA:\n",
    "    \n",
    "    def __init__(self,fileObj,nGenes):\n",
    "        \n",
    "        '''\n",
    "        Class Parameters:\n",
    "        \n",
    "        fileObj = csv file containing enriched genes at each arc, from ParetoTI \n",
    "        nGenes = number of genes to include for the enrichment analysis from each archetype\n",
    "        \n",
    "        '''\n",
    "        \n",
    "        self.fileObj = pd.read_csv(fileObj)\n",
    "        self.nGenes = nGenes\n",
    "        self.arcDict = {}\n",
    "        \n",
    "    def topGenesOnly(self):\n",
    "        \n",
    "        '''\n",
    "        Returns a dictionary of panda dataframes consisting top nGenes and the log mean wilcoxon score for each\n",
    "        archetype\n",
    "        \n",
    "        '''\n",
    "    \n",
    "        arches = self.fileObj.iloc[:, 0].unique()\n",
    "              \n",
    "        for arc in arches:\n",
    "            \n",
    "            eachArch = self.fileObj['arch_name'] == arc\n",
    "                        \n",
    "            topGenes = self.fileObj[eachArch][:self.nGenes]\n",
    "            \n",
    "            self.arcDict[arc] = pd.DataFrame(list(topGenes[\"mean_diff\"]),list(topGenes[\"genes\"]))\n",
    "            \n",
    "        return self.arcDict\n",
    "        \n",
    "    def runGSEA(self):\n",
    "        \n",
    "        '''Runs a pre-rank gene set enrichment analysis for each archetype'''\n",
    "        \n",
    "        for arc,data in self.arcDict.items():\n",
    "            \n",
    "            data.to_csv('rankfile')\n",
    "            \n",
    "            df = pd.read_csv('rankfile',header=None)[2:]\n",
    "            \n",
    "            results = gseapy.prerank(rnk = df,gene_sets='GO_Biological_Process_2018', processes=10,min_size = 1)\n",
    "            \n",
    "            print(arc)\n",
    "            print(results.res2d.sort_values('nes', ascending=False).head(10))\n",
    "            print('\\n\\n')\n",
    "            \n",
    "    def saveGeneRanks(self):\n",
    "        \n",
    "        '''Write gene and scores to a file for STRING network analysis'''\n",
    "        \n",
    "        mergeData = [ dt for dt in self.arcDict.values() ]\n",
    "        \n",
    "        mergeData = pd.concat(mergeData)\n",
    "        \n",
    "        mergeData.reset_index(level=0, inplace=True)\n",
    "\n",
    "        mergeData.columns = [\"protein\",\"log-fold-change\"]\n",
    "        \n",
    "        mergeData.to_csv(\"CD4Arcs.txt\", sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "archetype_1\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "positive regulation of protein insertion into m...  0.956522  2.658289   \n",
      "antimicrobial humoral immune response mediated ...  1.000000  2.650991   \n",
      "positive regulation of establishment of protein...  0.956522  2.573099   \n",
      "protein maturation (GO:0051604)                     0.818182  2.563989   \n",
      "dendritic cell chemotaxis (GO:0002407)              0.869565  2.526848   \n",
      "positive regulation of protein localization to ...  0.956522  2.522068   \n",
      "granulocyte chemotaxis (GO:0071621)                 0.869565  2.518257   \n",
      "positive regulation of mitochondrial outer memb...  0.956522  2.514861   \n",
      "regulation of protein insertion into mitochondr...  0.956522  2.502559   \n",
      "regulation of phosphatidylinositol 3-kinase sig...  0.869565  2.490040   \n",
      "\n",
      "                                                        pval  fdr  \\\n",
      "Term                                                                \n",
      "positive regulation of protein insertion into m...  0.165618  1.0   \n",
      "antimicrobial humoral immune response mediated ...  0.078629  1.0   \n",
      "positive regulation of establishment of protein...  0.169043  1.0   \n",
      "protein maturation (GO:0051604)                     0.115658  1.0   \n",
      "dendritic cell chemotaxis (GO:0002407)              0.321508  1.0   \n",
      "positive regulation of protein localization to ...  0.161290  1.0   \n",
      "granulocyte chemotaxis (GO:0071621)                 0.306695  1.0   \n",
      "positive regulation of mitochondrial outer memb...  0.165992  1.0   \n",
      "regulation of protein insertion into mitochondr...  0.179798  1.0   \n",
      "regulation of phosphatidylinositol 3-kinase sig...  0.336245  1.0   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "positive regulation of protein insertion into m...            31   \n",
      "antimicrobial humoral immune response mediated ...            50   \n",
      "positive regulation of establishment of protein...           111   \n",
      "protein maturation (GO:0051604)                               84   \n",
      "dendritic cell chemotaxis (GO:0002407)                        16   \n",
      "positive regulation of protein localization to ...            96   \n",
      "granulocyte chemotaxis (GO:0071621)                           57   \n",
      "positive regulation of mitochondrial outer memb...            38   \n",
      "regulation of protein insertion into mitochondr...            31   \n",
      "regulation of phosphatidylinositol 3-kinase sig...            81   \n",
      "\n",
      "                                                    matched_size      genes  \\\n",
      "Term                                                                          \n",
      "positive regulation of protein insertion into m...             1       GZMB   \n",
      "antimicrobial humoral immune response mediated ...             1       GNLY   \n",
      "positive regulation of establishment of protein...             1       GZMB   \n",
      "protein maturation (GO:0051604)                                2  GZMB;GZMH   \n",
      "dendritic cell chemotaxis (GO:0002407)                         1       CCL5   \n",
      "positive regulation of protein localization to ...             1       GZMB   \n",
      "granulocyte chemotaxis (GO:0071621)                            1       CCL5   \n",
      "positive regulation of mitochondrial outer memb...             1       GZMB   \n",
      "regulation of protein insertion into mitochondr...             1       GZMB   \n",
      "regulation of phosphatidylinositol 3-kinase sig...             1       CCL5   \n",
      "\n",
      "                                                   ledge_genes  \n",
      "Term                                                            \n",
      "positive regulation of protein insertion into m...        GZMB  \n",
      "antimicrobial humoral immune response mediated ...        GNLY  \n",
      "positive regulation of establishment of protein...        GZMB  \n",
      "protein maturation (GO:0051604)                      GZMB;GZMH  \n",
      "dendritic cell chemotaxis (GO:0002407)                    CCL5  \n",
      "positive regulation of protein localization to ...        GZMB  \n",
      "granulocyte chemotaxis (GO:0071621)                       CCL5  \n",
      "positive regulation of mitochondrial outer memb...        GZMB  \n",
      "regulation of protein insertion into mitochondr...        GZMB  \n",
      "regulation of phosphatidylinositol 3-kinase sig...        CCL5  \n",
      "\n",
      "\n",
      "\n",
      "archetype_2\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "platelet degranulation (GO:0002576)                 0.731969  2.818378   \n",
      "negative regulation of protein metabolic proces...  1.000000  2.759448   \n",
      "homotypic cell-cell adhesion (GO:0034109)           1.000000  2.758455   \n",
      "regulation of macrophage derived foam cell diff...  0.909091  2.757377   \n",
      "regulated exocytosis (GO:0045055)                   0.731969  2.754522   \n",
      "angiogenesis involved in wound healing (GO:0060...  1.000000  2.733864   \n",
      "positive regulation of peptidyl-tyrosine phosph...  1.000000  2.729970   \n",
      "regulation of megakaryocyte differentiation (GO...  0.863636  2.722317   \n",
      "transmembrane receptor protein tyrosine kinase ...  1.000000  2.710985   \n",
      "positive regulation of protein kinase activity ...  1.000000  2.702703   \n",
      "\n",
      "                                                        pval       fdr  \\\n",
      "Term                                                                     \n",
      "platelet degranulation (GO:0002576)                 0.003082  0.589958   \n",
      "negative regulation of protein metabolic proces...  0.068607  0.578725   \n",
      "homotypic cell-cell adhesion (GO:0034109)           0.093418  0.411538   \n",
      "regulation of macrophage derived foam cell diff...  0.041667  0.311095   \n",
      "regulated exocytosis (GO:0045055)                   0.012121  0.265286   \n",
      "angiogenesis involved in wound healing (GO:0060...  0.095634  0.268281   \n",
      "positive regulation of peptidyl-tyrosine phosph...  0.084016  0.241258   \n",
      "regulation of megakaryocyte differentiation (GO...  0.041367  0.224409   \n",
      "transmembrane receptor protein tyrosine kinase ...  0.103734  0.217816   \n",
      "positive regulation of protein kinase activity ...  0.076923  0.220844   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "platelet degranulation (GO:0002576)                          125   \n",
      "negative regulation of protein metabolic proces...            44   \n",
      "homotypic cell-cell adhesion (GO:0034109)                     39   \n",
      "regulation of macrophage derived foam cell diff...            30   \n",
      "regulated exocytosis (GO:0045055)                            149   \n",
      "angiogenesis involved in wound healing (GO:0060...             8   \n",
      "positive regulation of peptidyl-tyrosine phosph...           117   \n",
      "regulation of megakaryocyte differentiation (GO...            50   \n",
      "transmembrane receptor protein tyrosine kinase ...           397   \n",
      "positive regulation of protein kinase activity ...           173   \n",
      "\n",
      "                                                    matched_size  \\\n",
      "Term                                                               \n",
      "platelet degranulation (GO:0002576)                            6   \n",
      "negative regulation of protein metabolic proces...             1   \n",
      "homotypic cell-cell adhesion (GO:0034109)                      1   \n",
      "regulation of macrophage derived foam cell diff...             2   \n",
      "regulated exocytosis (GO:0045055)                              6   \n",
      "angiogenesis involved in wound healing (GO:0060...             1   \n",
      "positive regulation of peptidyl-tyrosine phosph...             1   \n",
      "regulation of megakaryocyte differentiation (GO...             2   \n",
      "transmembrane receptor protein tyrosine kinase ...             1   \n",
      "positive regulation of protein kinase activity ...             1   \n",
      "\n",
      "                                                                                 genes  \\\n",
      "Term                                                                                     \n",
      "platelet degranulation (GO:0002576)                 ITGB3;PPBP;PF4;ITGA2B;SPARC;RAB27B   \n",
      "negative regulation of protein metabolic proces...                               ITGB3   \n",
      "homotypic cell-cell adhesion (GO:0034109)                                        ITGB3   \n",
      "regulation of macrophage derived foam cell diff...                           ITGB3;PF4   \n",
      "regulated exocytosis (GO:0045055)                   ITGB3;PPBP;PF4;ITGA2B;SPARC;RAB27B   \n",
      "angiogenesis involved in wound healing (GO:0060...                               ITGB3   \n",
      "positive regulation of peptidyl-tyrosine phosph...                               ITGB3   \n",
      "regulation of megakaryocyte differentiation (GO...                          PF4;ITGA2B   \n",
      "transmembrane receptor protein tyrosine kinase ...                               ITGB3   \n",
      "positive regulation of protein kinase activity ...                               ITGB3   \n",
      "\n",
      "                                                                    ledge_genes  \n",
      "Term                                                                             \n",
      "platelet degranulation (GO:0002576)                 ITGB3;PPBP;PF4;ITGA2B;SPARC  \n",
      "negative regulation of protein metabolic proces...                        ITGB3  \n",
      "homotypic cell-cell adhesion (GO:0034109)                                 ITGB3  \n",
      "regulation of macrophage derived foam cell diff...                    ITGB3;PF4  \n",
      "regulated exocytosis (GO:0045055)                   ITGB3;PPBP;PF4;ITGA2B;SPARC  \n",
      "angiogenesis involved in wound healing (GO:0060...                        ITGB3  \n",
      "positive regulation of peptidyl-tyrosine phosph...                        ITGB3  \n",
      "regulation of megakaryocyte differentiation (GO...                   PF4;ITGA2B  \n",
      "transmembrane receptor protein tyrosine kinase ...                        ITGB3  \n",
      "positive regulation of protein kinase activity ...                        ITGB3  \n",
      "\n",
      "\n",
      "\n",
      "archetype_3\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "aminoglycan metabolic process (GO:0006022)          1.000000  2.982365   \n",
      "glycosaminoglycan metabolic process (GO:0030203)    1.000000  2.851122   \n",
      "inflammatory response (GO:0006954)                  0.578947  2.773325   \n",
      "sulfur compound biosynthetic process (GO:0044272)   1.000000  2.719962   \n",
      "dermatan sulfate metabolic process (GO:0030205)     1.000000  2.702703   \n",
      "monocarboxylic acid biosynthetic process (GO:00...  1.000000  2.702703   \n",
      "nervous system development (GO:0007399)             1.000000  2.690058   \n",
      "sulfur compound catabolic process (GO:0044273)      1.000000  2.663887   \n",
      "chondroitin sulfate biosynthetic process (GO:00...  1.000000  2.663887   \n",
      "superoxide anion generation (GO:0042554)            0.727273  2.655957   \n",
      "\n",
      "                                                        pval       fdr  \\\n",
      "Term                                                                     \n",
      "aminoglycan metabolic process (GO:0006022)          0.079727  1.000000   \n",
      "glycosaminoglycan metabolic process (GO:0030203)    0.096280  1.000000   \n",
      "inflammatory response (GO:0006954)                  0.072897  1.000000   \n",
      "sulfur compound biosynthetic process (GO:0044272)   0.091097  1.000000   \n",
      "dermatan sulfate metabolic process (GO:0030205)     0.085595  0.853971   \n",
      "monocarboxylic acid biosynthetic process (GO:00...  0.092784  0.853971   \n",
      "nervous system development (GO:0007399)             0.095238  0.801247   \n",
      "sulfur compound catabolic process (GO:0044273)      0.085714  0.752770   \n",
      "chondroitin sulfate biosynthetic process (GO:00...  0.065737  0.752770   \n",
      "superoxide anion generation (GO:0042554)            0.171779  0.719056   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "aminoglycan metabolic process (GO:0006022)                    31   \n",
      "glycosaminoglycan metabolic process (GO:0030203)              62   \n",
      "inflammatory response (GO:0006954)                           253   \n",
      "sulfur compound biosynthetic process (GO:0044272)            123   \n",
      "dermatan sulfate metabolic process (GO:0030205)               13   \n",
      "monocarboxylic acid biosynthetic process (GO:00...            69   \n",
      "nervous system development (GO:0007399)                      456   \n",
      "sulfur compound catabolic process (GO:0044273)                36   \n",
      "chondroitin sulfate biosynthetic process (GO:00...            20   \n",
      "superoxide anion generation (GO:0042554)                      10   \n",
      "\n",
      "                                                    matched_size  \\\n",
      "Term                                                               \n",
      "aminoglycan metabolic process (GO:0006022)                     1   \n",
      "glycosaminoglycan metabolic process (GO:0030203)               1   \n",
      "inflammatory response (GO:0006954)                             5   \n",
      "sulfur compound biosynthetic process (GO:0044272)              1   \n",
      "dermatan sulfate metabolic process (GO:0030205)                1   \n",
      "monocarboxylic acid biosynthetic process (GO:00...             1   \n",
      "nervous system development (GO:0007399)                        1   \n",
      "sulfur compound catabolic process (GO:0044273)                 1   \n",
      "chondroitin sulfate biosynthetic process (GO:00...             1   \n",
      "superoxide anion generation (GO:0042554)                       2   \n",
      "\n",
      "                                                                          genes  \\\n",
      "Term                                                                              \n",
      "aminoglycan metabolic process (GO:0006022)                                 VCAN   \n",
      "glycosaminoglycan metabolic process (GO:0030203)                           VCAN   \n",
      "inflammatory response (GO:0006954)                  FPR1;CYBB;S100A9;LYZ;S100A8   \n",
      "sulfur compound biosynthetic process (GO:0044272)                          VCAN   \n",
      "dermatan sulfate metabolic process (GO:0030205)                            VCAN   \n",
      "monocarboxylic acid biosynthetic process (GO:00...                         VCAN   \n",
      "nervous system development (GO:0007399)                                    VCAN   \n",
      "sulfur compound catabolic process (GO:0044273)                             VCAN   \n",
      "chondroitin sulfate biosynthetic process (GO:00...                         VCAN   \n",
      "superoxide anion generation (GO:0042554)                              NCF2;CYBB   \n",
      "\n",
      "                                                                    ledge_genes  \n",
      "Term                                                                             \n",
      "aminoglycan metabolic process (GO:0006022)                                 VCAN  \n",
      "glycosaminoglycan metabolic process (GO:0030203)                           VCAN  \n",
      "inflammatory response (GO:0006954)                  FPR1;CYBB;S100A9;LYZ;S100A8  \n",
      "sulfur compound biosynthetic process (GO:0044272)                          VCAN  \n",
      "dermatan sulfate metabolic process (GO:0030205)                            VCAN  \n",
      "monocarboxylic acid biosynthetic process (GO:00...                         VCAN  \n",
      "nervous system development (GO:0007399)                                    VCAN  \n",
      "sulfur compound catabolic process (GO:0044273)                             VCAN  \n",
      "chondroitin sulfate biosynthetic process (GO:00...                         VCAN  \n",
      "superoxide anion generation (GO:0042554)                              NCF2;CYBB  \n",
      "\n",
      "\n",
      "\n",
      "archetype_4\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "transmembrane receptor protein tyrosine kinase ...  0.818182  2.770713   \n",
      "negative regulation of G0 to G1 transition (GO:...  1.000000  2.709069   \n",
      "nucleotide biosynthetic process (GO:0009165)        1.000000  2.663887   \n",
      "mitotic spindle organization (GO:0007052)           0.650000  2.662486   \n",
      "negative regulation of chromosome organization ...  0.913043  2.573529   \n",
      "negative regulation of cell cycle process (GO:0...  1.000000  2.569545   \n",
      "regulation of G0 to G1 transition (GO:0070316)      1.000000  2.544248   \n",
      "regulation of apoptotic process (GO:0042981)        0.913043  2.519194   \n",
      "carbohydrate derivative biosynthetic process (G...  1.000000  2.504355   \n",
      "positive regulation of programmed cell death (G...  0.913043  2.502085   \n",
      "\n",
      "                                                        pval  fdr  \\\n",
      "Term                                                                \n",
      "transmembrane receptor protein tyrosine kinase ...  0.070746  1.0   \n",
      "negative regulation of G0 to G1 transition (GO:...  0.088050  1.0   \n",
      "nucleotide biosynthetic process (GO:0009165)        0.068687  1.0   \n",
      "mitotic spindle organization (GO:0007052)           0.062167  1.0   \n",
      "negative regulation of chromosome organization ...  0.243011  1.0   \n",
      "negative regulation of cell cycle process (GO:0...  0.078278  1.0   \n",
      "regulation of G0 to G1 transition (GO:0070316)      0.085437  1.0   \n",
      "regulation of apoptotic process (GO:0042981)        0.239583  1.0   \n",
      "carbohydrate derivative biosynthetic process (G...  0.080306  1.0   \n",
      "positive regulation of programmed cell death (G...  0.217391  1.0   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "transmembrane receptor protein tyrosine kinase ...           397   \n",
      "negative regulation of G0 to G1 transition (GO:...            38   \n",
      "nucleotide biosynthetic process (GO:0009165)                  27   \n",
      "mitotic spindle organization (GO:0007052)                     75   \n",
      "negative regulation of chromosome organization ...             6   \n",
      "negative regulation of cell cycle process (GO:0...            82   \n",
      "regulation of G0 to G1 transition (GO:0070316)                40   \n",
      "regulation of apoptotic process (GO:0042981)                 816   \n",
      "carbohydrate derivative biosynthetic process (G...            30   \n",
      "positive regulation of programmed cell death (G...           258   \n",
      "\n",
      "                                                    matched_size  \\\n",
      "Term                                                               \n",
      "transmembrane receptor protein tyrosine kinase ...             2   \n",
      "negative regulation of G0 to G1 transition (GO:...             1   \n",
      "nucleotide biosynthetic process (GO:0009165)                   1   \n",
      "mitotic spindle organization (GO:0007052)                      4   \n",
      "negative regulation of chromosome organization ...             1   \n",
      "negative regulation of cell cycle process (GO:0...             1   \n",
      "regulation of G0 to G1 transition (GO:0070316)                 1   \n",
      "regulation of apoptotic process (GO:0042981)                   1   \n",
      "carbohydrate derivative biosynthetic process (G...             1   \n",
      "positive regulation of programmed cell death (G...             1   \n",
      "\n",
      "                                                                     genes  \\\n",
      "Term                                                                         \n",
      "transmembrane receptor protein tyrosine kinase ...            STMN1;SHCBP1   \n",
      "negative regulation of G0 to G1 transition (GO:...                    RRM2   \n",
      "nucleotide biosynthetic process (GO:0009165)                          RRM2   \n",
      "mitotic spindle organization (GO:0007052)           STMN1;MYBL2;TPX2;KIF11   \n",
      "negative regulation of chromosome organization ...                   TOP2A   \n",
      "negative regulation of cell cycle process (GO:0...                    RRM2   \n",
      "regulation of G0 to G1 transition (GO:0070316)                        RRM2   \n",
      "regulation of apoptotic process (GO:0042981)                         TOP2A   \n",
      "carbohydrate derivative biosynthetic process (G...                    RRM2   \n",
      "positive regulation of programmed cell death (G...                   TOP2A   \n",
      "\n",
      "                                                               ledge_genes  \n",
      "Term                                                                        \n",
      "transmembrane receptor protein tyrosine kinase ...            STMN1;SHCBP1  \n",
      "negative regulation of G0 to G1 transition (GO:...                    RRM2  \n",
      "nucleotide biosynthetic process (GO:0009165)                          RRM2  \n",
      "mitotic spindle organization (GO:0007052)           STMN1;MYBL2;TPX2;KIF11  \n",
      "negative regulation of chromosome organization ...                   TOP2A  \n",
      "negative regulation of cell cycle process (GO:0...                    RRM2  \n",
      "regulation of G0 to G1 transition (GO:0070316)                        RRM2  \n",
      "regulation of apoptotic process (GO:0042981)                         TOP2A  \n",
      "carbohydrate derivative biosynthetic process (G...                    RRM2  \n",
      "positive regulation of programmed cell death (G...                   TOP2A  \n",
      "\n",
      "\n",
      "\n",
      "archetype_5\n",
      "                                                     es       nes      pval  \\\n",
      "Term                                                                          \n",
      "cellular response to organic substance (GO:0071...  1.0  2.839490  0.004950   \n",
      "negative regulation of cytokine production (GO:...  1.0  2.809332  0.082278   \n",
      "negative regulation of DNA binding (GO:0043392)     1.0  2.787879  0.084388   \n",
      "Wnt signaling pathway, calcium modulating pathw...  1.0  2.787541  0.086316   \n",
      "striated muscle cell development (GO:0055002)       1.0  2.764091  0.057851   \n",
      "epithelial to mesenchymal transition (GO:0001837)   1.0  2.760110  0.090526   \n",
      "radial glial cell differentiation (GO:0060019)      1.0  2.745941  0.083333   \n",
      "negative regulation of gene expression (GO:0010...  1.0  2.737118  0.067762   \n",
      "alpha-beta T cell activation (GO:0046631)           1.0  2.734839  0.093361   \n",
      "mammary gland development (GO:0030879)              1.0  2.730619  0.074844   \n",
      "\n",
      "                                                         fdr  geneset_size  \\\n",
      "Term                                                                         \n",
      "cellular response to organic substance (GO:0071...  0.430040           134   \n",
      "negative regulation of cytokine production (GO:...  0.352394           111   \n",
      "negative regulation of DNA binding (GO:0043392)     0.321082            35   \n",
      "Wnt signaling pathway, calcium modulating pathw...  0.252486            37   \n",
      "striated muscle cell development (GO:0055002)       0.377697            20   \n",
      "epithelial to mesenchymal transition (GO:0001837)   0.344611            38   \n",
      "radial glial cell differentiation (GO:0060019)      0.374966             7   \n",
      "negative regulation of gene expression (GO:0010...  0.370177           619   \n",
      "alpha-beta T cell activation (GO:0046631)           0.358487             6   \n",
      "mammary gland development (GO:0030879)              0.337408            25   \n",
      "\n",
      "                                                    matched_size      genes  \\\n",
      "Term                                                                          \n",
      "cellular response to organic substance (GO:0071...             2  LEF1;CCR7   \n",
      "negative regulation of cytokine production (GO:...             1       LEF1   \n",
      "negative regulation of DNA binding (GO:0043392)                1       LEF1   \n",
      "Wnt signaling pathway, calcium modulating pathw...             1       LEF1   \n",
      "striated muscle cell development (GO:0055002)                  1       LEF1   \n",
      "epithelial to mesenchymal transition (GO:0001837)              1       LEF1   \n",
      "radial glial cell differentiation (GO:0060019)                 1       LEF1   \n",
      "negative regulation of gene expression (GO:0010...             1       LEF1   \n",
      "alpha-beta T cell activation (GO:0046631)                      1       LEF1   \n",
      "mammary gland development (GO:0030879)                         1       LEF1   \n",
      "\n",
      "                                                   ledge_genes  \n",
      "Term                                                            \n",
      "cellular response to organic substance (GO:0071...   LEF1;CCR7  \n",
      "negative regulation of cytokine production (GO:...        LEF1  \n",
      "negative regulation of DNA binding (GO:0043392)           LEF1  \n",
      "Wnt signaling pathway, calcium modulating pathw...        LEF1  \n",
      "striated muscle cell development (GO:0055002)             LEF1  \n",
      "epithelial to mesenchymal transition (GO:0001837)         LEF1  \n",
      "radial glial cell differentiation (GO:0060019)            LEF1  \n",
      "negative regulation of gene expression (GO:0010...        LEF1  \n",
      "alpha-beta T cell activation (GO:0046631)                 LEF1  \n",
      "mammary gland development (GO:0030879)                    LEF1  \n",
      "\n",
      "\n",
      "\n",
      "archetype_6\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "regulation of protein processing (GO:0070613)       0.663987  2.865723   \n",
      "Fc-gamma receptor signaling pathway involved in...  0.663987  2.830760   \n",
      "Fc receptor mediated stimulatory signaling path...  0.663987  2.821523   \n",
      "regulation of protein activation cascade (GO:20...  0.663987  2.803745   \n",
      "integrin activation (GO:0033622)                    1.000000  2.775097   \n",
      "regulation of immune effector process (GO:0002697)  0.663987  2.764887   \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)    0.663987  2.748908   \n",
      "regulation of acute inflammatory response (GO:0...  0.663987  2.745863   \n",
      "regulation of humoral immune response (GO:0002920)  0.663987  2.704641   \n",
      "regulation of glucose import in response to ins...  1.000000  2.701750   \n",
      "\n",
      "                                                        pval       fdr  \\\n",
      "Term                                                                     \n",
      "regulation of protein processing (GO:0070613)       0.044248  0.757530   \n",
      "Fc-gamma receptor signaling pathway involved in...  0.042735  0.435044   \n",
      "Fc receptor mediated stimulatory signaling path...  0.037543  0.306903   \n",
      "regulation of protein activation cascade (GO:20...  0.030822  0.244470   \n",
      "integrin activation (GO:0033622)                    0.083857  0.241076   \n",
      "regulation of immune effector process (GO:0002697)  0.044068  0.211219   \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)    0.042159  0.204697   \n",
      "regulation of acute inflammatory response (GO:0...  0.032680  0.182311   \n",
      "regulation of humoral immune response (GO:0002920)  0.042975  0.211550   \n",
      "regulation of glucose import in response to ins...  0.092593  0.194445   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "regulation of protein processing (GO:0070613)                129   \n",
      "Fc-gamma receptor signaling pathway involved in...           134   \n",
      "Fc receptor mediated stimulatory signaling path...           136   \n",
      "regulation of protein activation cascade (GO:20...           109   \n",
      "integrin activation (GO:0033622)                               9   \n",
      "regulation of immune effector process (GO:0002697)           115   \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)             135   \n",
      "regulation of acute inflammatory response (GO:0...           122   \n",
      "regulation of humoral immune response (GO:0002920)           114   \n",
      "regulation of glucose import in response to ins...            13   \n",
      "\n",
      "                                                    matched_size  \\\n",
      "Term                                                               \n",
      "regulation of protein processing (GO:0070613)                  5   \n",
      "Fc-gamma receptor signaling pathway involved in...             5   \n",
      "Fc receptor mediated stimulatory signaling path...             5   \n",
      "regulation of protein activation cascade (GO:20...             5   \n",
      "integrin activation (GO:0033622)                               1   \n",
      "regulation of immune effector process (GO:0002697)             5   \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)               5   \n",
      "regulation of acute inflammatory response (GO:0...             5   \n",
      "regulation of humoral immune response (GO:0002920)             5   \n",
      "regulation of glucose import in response to ins...             1   \n",
      "\n",
      "                                                                           genes  \\\n",
      "Term                                                                               \n",
      "regulation of protein processing (GO:0070613)       IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "Fc-gamma receptor signaling pathway involved in...  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "Fc receptor mediated stimulatory signaling path...  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "regulation of protein activation cascade (GO:20...  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "integrin activation (GO:0033622)                                            MZB1   \n",
      "regulation of immune effector process (GO:0002697)  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)    IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "regulation of acute inflammatory response (GO:0...  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "regulation of humoral immune response (GO:0002920)  IGHG3;IGHG1;IGHG4;IGHG2;IGKC   \n",
      "regulation of glucose import in response to ins...                          MZB1   \n",
      "\n",
      "                                                                ledge_genes  \n",
      "Term                                                                         \n",
      "regulation of protein processing (GO:0070613)       IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "Fc-gamma receptor signaling pathway involved in...  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "Fc receptor mediated stimulatory signaling path...  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "regulation of protein activation cascade (GO:20...  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "integrin activation (GO:0033622)                                       MZB1  \n",
      "regulation of immune effector process (GO:0002697)  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "Fc-gamma receptor signaling pathway (GO:0038094)    IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "regulation of acute inflammatory response (GO:0...  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "regulation of humoral immune response (GO:0002920)  IGHG3;IGHG1;IGHG4;IGHG2  \n",
      "regulation of glucose import in response to ins...                     MZB1  \n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''Run a pre-ranked gene set enrichment analysis on the results of a Pareto task inference done on all PBMCs.'''\n",
    "gsea = ParetoGSEA(fileObj = 'covidArcs.csv',nGenes = 25)\n",
    "gsea.topGenesOnly()\n",
    "gsea.runGSEA()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STRING to create protein-protein interactions at each Archetype\n",
    "\n",
    "Key biological processes occuring at each archetype will be determined by STRING's GO biological processes. I use a .80 interaction score and remove nodes that are disconnected. The networks are displayed by confidence seen by the strength(thickness and color) of the edge connecting nodes.\n",
    "\n",
    "Interior node colors(NOT THE OUTER SHADE) are labeled for processes/function mentioned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Create a file with gene and scores to be used for a network analysis using STRING online tool.'''\n",
    "\n",
    "gsea = ParetoGSEA(fileObj = 'covidArcs.csv',nGenes = 25)\n",
    "gsea.topGenesOnly()\n",
    "gsea.saveGeneRanks()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 1:\n",
    "\n",
    "The main processes include immune and cellular defense response by granzymes facilitated by Perforin-1(PRF1). IL2RB also likely induces T cell signaling here.\n",
    "\n",
    "Different from the expected defense responses, SYNE1, SYNE2 and CX3CR1 may contribute to dendritic cell chemotaxis and cerebral cortex migration(red) through actin filament binding(blue). These processes suggest brain cell protrusion by SarsCov2 and the ability for neuropathology. Drugs to inhibit premature formation of dendritic cells and actin filament polymerization may be necessary in the future as these would slow virus spread within the brain.\n",
    "\n",
    "<img src=\"arc1.png\" width=\"800\" />"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 2\n",
    "\n",
    "ITGA2B and ITGB3 interaction brings about rapid platelet aggregation physically plugging rupturing endothelial cell surfaces.(STRING). HIST1H2AC role in transcriptional regulation and therefore gene activity may induce integrin family proteins at the site of the extracellular matrix to result in clotting effects; these events would suggest gene expression responses due to blood vessel damage. PF4, SPARC, and PF4V1 contribute to wound healing(blue) based on platelet activation(red) and contribute to platelet aggregation resulting in blood clots. These gene associations contribute to the vascular damage, given past research on SarsCov ability to enter blood vessel for widespread infection, SarsCov2 may not only have the same properties, but may cause microvascular damage in the process leading to blood coagulation(1). Inhibition of ITGA2B and ITGB3 to preven rapid coagulation through blood thinners may benefit infected patients with cardiovascular disease. \n",
    "\n",
    "1. https://www.frontiersin.org/articles/10.3389/fphar.2020.00836/full\n",
    "\n",
    "<img src=\"arc2.png\" width=\"800\" />"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 3\n",
    "\n",
    "Genes in response to coagulation include SERPINA1,CLU,and VCAN. These genes relate to leukocyte differentiation, myeloid leukocyte activation, and white blood cell mediated immunity evident through CD molecule function, macrophage proteins, and neutrophil activation. \n",
    "\n",
    "S100A9, S100A8, and S100A12 contribute to inflammatory responses not only for leukocytes but for cytokines and chemokines and induce neutrophil chemotaxis through the ERK1/2 and MAPK dependent processes; these signaling processes stimulate inflammotory response and thus viral role can be assumed to impact these signaling pathways. Inhibiting ERK1 and MAPK pathways in cardiovascular patients may lessen overall aggravation to prevent an overdrive of a cytokine production.\n",
    "\n",
    "<img src=\"arc3.png\" width=\"800\" />"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 4\n",
    "\n",
    "Genes at this arch contribute to mitotic cell cycle processes,cell division,and other cell cycle processes. Certain genes contribute to cell differentiation and growth including SHCBP1, an activator of fibroblast growth factor in neuronal progenitor cells and MYBL2. There may be significance in the CCNA2 to induce cyclin dependent reactions in cell differentiation. Furthermore, based on further research conducted by UCSF to study SarsCov2 dependency on kinases, the pathways may not suggest growth, but rather disfunction in these growth pathways. Further analysis of UCSFs findings by European Molecular Biology Laboratoryâ€™s European Biosciences Institute found 49 kinases with abnormal function including a network involving p38/MAPK pathways leading to the production of inflammation-inducing cytokines resulting in immune system overdrive(2). However, UCSF's findings were not reflective within this archetype's network, therefore, it is simply implicated that SarsCov2 has an involvment with cyclins for its own use for viral replication or cell takeover.\n",
    "\n",
    "2. https://www.ucsf.edu/news/2020/06/417936/covid-19-relies-cells-master-regulators-survival\n",
    "\n",
    "<img src=\"arc4.png\" width=\"800\" />"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 5\n",
    "\n",
    "LEF1 induced Wnt signaling activates transcription of epithelial to mesenchymal transition(EMT)(3) involving cell migration, metastasis, and invasion. However, this proclaimed biomarker has only been suggested among cancerous cells. Similar EMT pathways may be suggested in SarsCov2. MYC activates growth related genes and binds to the VEGFA promoter promoting VEGFA production and sprouting angiogenesis, thus allowing permeability in the blood vessel. Given BCL2's function to inhibit apoptosis in nueral cells and lymphocytes, the pathway involving LEF1, MYC, and BCL2 suggest vascular endothelial cell growth resulting in the permeability of infected lymphocytes through blood vessels. Thus, looking into VEGFA and BCL2 inhibitors may combat virus spread in blood vessels. \n",
    "\n",
    "The link between LTB and BIRC3 also suggest metastatic properties thus providing further details on cell migration induced by the LTB cell membrane receptor to call in an inflammotory response that may affect a modulator of inflammotory signaling,BIRC3, which may affect cell invasion and metastasis.\n",
    "\n",
    "\n",
    "3. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5489786/\n",
    "\n",
    "<img src=\"arc5.png\" width=\"800\" />"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archetype 6\n",
    "\n",
    "Genes in this archetype involve humoral responses and B cell antigen recoginition, and B-cell differentiation including MZB1,IGF,TNFRSF17, and POU2AF1.\n",
    "\n",
    "<img src=\"arc6.png\" width=\"800\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Multiple archetypes suggest neuronal differentiation from archetype 1, blood clotting from archetype 2, and invasive cell properties from archetype 5. To further analyze archetype 5 that details cell invasion and metastasis pathways, I will perform another Pareto analysis on the cell type distinguished by archetype 5 only. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pareto Archetypal Analysis among CD4 T cells identified at Archetype 5\n",
    "\n",
    "# R code shown below. \n",
    "\n",
    "nearCells = bin_cells_by_arch(mergeDat$data,mergeDat$arc_col,bin_prop = 0.1,return_names = TRUE,dist_cutoff = 0.01)\n",
    "\n",
    "celltypes <- data.frame(nearCells)\n",
    "\n",
    "celltypes <- data.frame(covid$cell.type)\n",
    "\n",
    "celltypes <- data.frame(cells = rownames(celltypes),celltypes,row.names = NULL)\n",
    "\n",
    "arc5 <- subset(celltypes,cells %in% nearCells$archetype_5)\n",
    "\n",
    "Idents(object = covid) <- \"cell.type\"\n",
    "\n",
    "indCells <- subset(covid,idents = c(\"CD4m T\",\"CD4n T\"))\n",
    "\n",
    "indCellsPCS <- data.frame(indCells@reductions$pca\\@cell.embeddings)\n",
    "\n",
    "indCellsPCS <- indCellsPCS[c(1:10)]\n",
    "\n",
    "indCellsPCS <- data.matrix(indCellsPCS)\n",
    "\n",
    "indCellsPCS <- t(indCellsPCS)\n",
    "\n",
    "indCellPcha <- k_fit_pch(indCellsPCS,ks = 3:8,maxiter = 500,seed=345)\n",
    "\n",
    "CD4PCHA <- fit_pch(indCellsPCS, noc = 3,conv_crit= 1e-06,maxiter = 500,volume_ratio = \"t_ratio\")\n",
    "\n",
    "plt_arc <- plot_arc(arc_data = CD4PCHA, data = indCellsPCS)\n",
    "\n",
    "print(plt_arc)\n",
    "\n",
    "<img src=\"CD4Arcs.png\" width=\"800\" />\n",
    "\n",
    "\n",
    "obj <- merge_arch_dist(arc_data = CD4PCHA,data = indCellsPCS,dist_metric = c(\"euclidean\",\"arch_weights\")[1],feature_data = indCells@assays$RNA\\@scale.data)\n",
    "\n",
    "CD4Genes <- find_decreasing_wilcox(obj\\\\$data,obj\\\\$arc_col,features = obj$features_col,bin = 0.1,method = \"BioQC\")\n",
    "\n",
    "CD4Stats <- get_top_decreasing(summary_genes = CD4Genes,summary_sets = NULL, cutoff_genes = 0.05,cutoff_metric = \"wilcoxon_p_val\",p.adjust.method = \"BH\")\n",
    "\n",
    "dt2 <- fwrite(x = CD4Stats$enriched_genes,file=\"CD4.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "archetype_1\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "regulation of cell proliferation (GO:0042127)       0.863636  2.908140   \n",
      "regulation of apoptotic process (GO:0042981)        0.698462  2.822348   \n",
      "negative regulation of binding (GO:0051100)         1.000000  2.778113   \n",
      "regulation of DNA binding (GO:0051101)              1.000000  2.736466   \n",
      "negative regulation of oxidoreductase activity ...  1.000000  2.727704   \n",
      "regulation of endodeoxyribonuclease activity (G...  1.000000  2.707794   \n",
      "negative regulation of interleukin-12 productio...  0.956522  2.686859   \n",
      "cellular protein catabolic process (GO:0044257)     1.000000  2.668523   \n",
      "positive regulation of MAPK cascade (GO:0043410)    0.818182  2.668099   \n",
      "regulation of T cell activation (GO:0050863)        0.818182  2.637292   \n",
      "\n",
      "                                                        pval       fdr  \\\n",
      "Term                                                                     \n",
      "regulation of cell proliferation (GO:0042127)       0.046693  1.000000   \n",
      "regulation of apoptotic process (GO:0042981)        0.049209  1.000000   \n",
      "negative regulation of binding (GO:0051100)         0.069767  0.932710   \n",
      "regulation of DNA binding (GO:0051101)              0.075000  0.999287   \n",
      "negative regulation of oxidoreductase activity ...  0.107822  0.872844   \n",
      "regulation of endodeoxyribonuclease activity (G...  0.096509  0.898205   \n",
      "negative regulation of interleukin-12 productio...  0.181624  0.912690   \n",
      "cellular protein catabolic process (GO:0044257)     0.089613  0.981981   \n",
      "positive regulation of MAPK cascade (GO:0043410)    0.050373  0.879929   \n",
      "regulation of T cell activation (GO:0050863)        0.099083  1.000000   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "regulation of cell proliferation (GO:0042127)                741   \n",
      "regulation of apoptotic process (GO:0042981)                 816   \n",
      "negative regulation of binding (GO:0051100)                   83   \n",
      "regulation of DNA binding (GO:0051101)                        53   \n",
      "negative regulation of oxidoreductase activity ...            12   \n",
      "regulation of endodeoxyribonuclease activity (G...             8   \n",
      "negative regulation of interleukin-12 productio...            12   \n",
      "cellular protein catabolic process (GO:0044257)               63   \n",
      "positive regulation of MAPK cascade (GO:0043410)             290   \n",
      "regulation of T cell activation (GO:0050863)                  45   \n",
      "\n",
      "                                                    matched_size  \\\n",
      "Term                                                               \n",
      "regulation of cell proliferation (GO:0042127)                  2   \n",
      "regulation of apoptotic process (GO:0042981)                   4   \n",
      "negative regulation of binding (GO:0051100)                    1   \n",
      "regulation of DNA binding (GO:0051101)                         1   \n",
      "negative regulation of oxidoreductase activity ...             1   \n",
      "regulation of endodeoxyribonuclease activity (G...             1   \n",
      "negative regulation of interleukin-12 productio...             1   \n",
      "cellular protein catabolic process (GO:0044257)                1   \n",
      "positive regulation of MAPK cascade (GO:0043410)               2   \n",
      "regulation of T cell activation (GO:0050863)                   2   \n",
      "\n",
      "                                                                        genes  \\\n",
      "Term                                                                            \n",
      "regulation of cell proliferation (GO:0042127)                  TNFRSF1B;IFIT3   \n",
      "regulation of apoptotic process (GO:0042981)        GZMA;TNFRSF1B;IFIT3;ITGB1   \n",
      "negative regulation of binding (GO:0051100)                              GZMA   \n",
      "regulation of DNA binding (GO:0051101)                                   GZMA   \n",
      "negative regulation of oxidoreductase activity ...                       GZMA   \n",
      "regulation of endodeoxyribonuclease activity (G...                       GZMA   \n",
      "negative regulation of interleukin-12 productio...                      TIGIT   \n",
      "cellular protein catabolic process (GO:0044257)                          GZMA   \n",
      "positive regulation of MAPK cascade (GO:0043410)                TNFRSF1B;CCL5   \n",
      "regulation of T cell activation (GO:0050863)                       TIGIT;CCL5   \n",
      "\n",
      "                                                            ledge_genes  \n",
      "Term                                                                     \n",
      "regulation of cell proliferation (GO:0042127)            TNFRSF1B;IFIT3  \n",
      "regulation of apoptotic process (GO:0042981)        GZMA;TNFRSF1B;IFIT3  \n",
      "negative regulation of binding (GO:0051100)                        GZMA  \n",
      "regulation of DNA binding (GO:0051101)                             GZMA  \n",
      "negative regulation of oxidoreductase activity ...                 GZMA  \n",
      "regulation of endodeoxyribonuclease activity (G...                 GZMA  \n",
      "negative regulation of interleukin-12 productio...                TIGIT  \n",
      "cellular protein catabolic process (GO:0044257)                    GZMA  \n",
      "positive regulation of MAPK cascade (GO:0043410)          TNFRSF1B;CCL5  \n",
      "regulation of T cell activation (GO:0050863)                 TIGIT;CCL5  \n",
      "\n",
      "\n",
      "\n",
      "archetype_2\n",
      "                                                          es       nes  \\\n",
      "Term                                                                     \n",
      "cell junction assembly (GO:0034329)                 0.863636  2.995649   \n",
      "epithelial cell differentiation (GO:0030855)        1.000000  2.777107   \n",
      "positive regulation of macromolecule metabolic ...  0.863636  2.710048   \n",
      "positive regulation of gene expression (GO:0010...  0.863636  2.669947   \n",
      "lens fiber cell differentiation (GO:0070306)        1.000000  2.668523   \n",
      "positive regulation of nuclear division (GO:005...  0.956522  2.638522   \n",
      "positive regulation of cellular biosynthetic pr...  0.956522  2.626239   \n",
      "regulation of interleukin-10 production (GO:003...  0.956522  2.625298   \n",
      "positive regulation of cytokine production (GO:...  0.956522  2.607253   \n",
      "positive regulation of cellular protein metabol...  0.956522  2.605709   \n",
      "\n",
      "                                                        pval       fdr  \\\n",
      "Term                                                                     \n",
      "cell junction assembly (GO:0034329)                 0.051081  0.386091   \n",
      "epithelial cell differentiation (GO:0030855)        0.073996  0.954278   \n",
      "positive regulation of macromolecule metabolic ...  0.044092  1.000000   \n",
      "positive regulation of gene expression (GO:0010...  0.064865  1.000000   \n",
      "lens fiber cell differentiation (GO:0070306)        0.089249  1.000000   \n",
      "positive regulation of nuclear division (GO:005...  0.183716  1.000000   \n",
      "positive regulation of cellular biosynthetic pr...  0.168750  1.000000   \n",
      "regulation of interleukin-10 production (GO:003...  0.139293  1.000000   \n",
      "positive regulation of cytokine production (GO:...  0.157787  1.000000   \n",
      "positive regulation of cellular protein metabol...  0.160083  1.000000   \n",
      "\n",
      "                                                    geneset_size  \\\n",
      "Term                                                               \n",
      "cell junction assembly (GO:0034329)                           60   \n",
      "epithelial cell differentiation (GO:0030855)                  97   \n",
      "positive regulation of macromolecule metabolic ...           277   \n",
      "positive regulation of gene expression (GO:0010...           772   \n",
      "lens fiber cell differentiation (GO:0070306)                   7   \n",
      "positive regulation of nuclear division (GO:005...            26   \n",
      "positive regulation of cellular biosynthetic pr...           160   \n",
      "regulation of interleukin-10 production (GO:003...            26   \n",
      "positive regulation of cytokine production (GO:...           221   \n",
      "positive regulation of cellular protein metabol...            92   \n",
      "\n",
      "                                                    matched_size        genes  \\\n",
      "Term                                                                            \n",
      "cell junction assembly (GO:0034329)                            2  ACTG1;LIMS1   \n",
      "epithelial cell differentiation (GO:0030855)                   1          MAF   \n",
      "positive regulation of macromolecule metabolic ...             2   CD28;LIMS1   \n",
      "positive regulation of gene expression (GO:0010...             2   CD28;LIMS1   \n",
      "lens fiber cell differentiation (GO:0070306)                   1          MAF   \n",
      "positive regulation of nuclear division (GO:005...             1         CD28   \n",
      "positive regulation of cellular biosynthetic pr...             1         CD28   \n",
      "regulation of interleukin-10 production (GO:003...             1         CD28   \n",
      "positive regulation of cytokine production (GO:...             1         CD28   \n",
      "positive regulation of cellular protein metabol...             1         CD28   \n",
      "\n",
      "                                                    ledge_genes  \n",
      "Term                                                             \n",
      "cell junction assembly (GO:0034329)                 ACTG1;LIMS1  \n",
      "epithelial cell differentiation (GO:0030855)                MAF  \n",
      "positive regulation of macromolecule metabolic ...   CD28;LIMS1  \n",
      "positive regulation of gene expression (GO:0010...   CD28;LIMS1  \n",
      "lens fiber cell differentiation (GO:0070306)                MAF  \n",
      "positive regulation of nuclear division (GO:005...         CD28  \n",
      "positive regulation of cellular biosynthetic pr...         CD28  \n",
      "regulation of interleukin-10 production (GO:003...         CD28  \n",
      "positive regulation of cytokine production (GO:...         CD28  \n",
      "positive regulation of cellular protein metabol...         CD28  \n",
      "\n",
      "\n",
      "\n",
      "archetype_3\n",
      "                                                     es       nes      pval  \\\n",
      "Term                                                                          \n",
      "regulation of interleukin-1 secretion (GO:0050704)  1.0  3.085467  0.154217   \n",
      "positive regulation of ERK1 and ERK2 cascade (G...  1.0  2.927400  0.212617   \n",
      "regulation of pseudopodium assembly (GO:0031272)    1.0  2.923122  0.170115   \n",
      "positive regulation of plasma membrane bounded ...  1.0  2.892682  0.198157   \n",
      "regulation of plasma membrane bounded cell proj...  1.0  2.877698  0.176201   \n",
      "calcium ion transmembrane import into cytosol (...  1.0  2.873563  0.175000   \n",
      "regulation of phosphatidylinositol 3-kinase act...  1.0  2.869440  0.191244   \n",
      "negative regulation of sequestering of calcium ...  1.0  2.862869  0.205011   \n",
      "negative regulation of dendritic cell apoptotic...  1.0  2.862049  0.191781   \n",
      "immunological synapse formation (GO:0001771)        1.0  2.859594  0.207763   \n",
      "\n",
      "                                                         fdr  geneset_size  \\\n",
      "Term                                                                         \n",
      "regulation of interleukin-1 secretion (GO:0050704)  0.170780             9   \n",
      "positive regulation of ERK1 and ERK2 cascade (G...  0.602138           202   \n",
      "regulation of pseudopodium assembly (GO:0031272)    0.428603            12   \n",
      "positive regulation of plasma membrane bounded ...  0.738487            70   \n",
      "regulation of plasma membrane bounded cell proj...  0.736228            40   \n",
      "calcium ion transmembrane import into cytosol (...  0.675592            37   \n",
      "regulation of phosphatidylinositol 3-kinase act...  0.647863            34   \n",
      "negative regulation of sequestering of calcium ...  0.618390            25   \n",
      "negative regulation of dendritic cell apoptotic...  0.573063             7   \n",
      "immunological synapse formation (GO:0001771)        0.537132             6   \n",
      "\n",
      "                                                    matched_size genes  \\\n",
      "Term                                                                     \n",
      "regulation of interleukin-1 secretion (GO:0050704)             1  CCR7   \n",
      "positive regulation of ERK1 and ERK2 cascade (G...             1  CCR7   \n",
      "regulation of pseudopodium assembly (GO:0031272)               1  CCR7   \n",
      "positive regulation of plasma membrane bounded ...             1  CCR7   \n",
      "regulation of plasma membrane bounded cell proj...             1  CCR7   \n",
      "calcium ion transmembrane import into cytosol (...             1  CCR7   \n",
      "regulation of phosphatidylinositol 3-kinase act...             1  CCR7   \n",
      "negative regulation of sequestering of calcium ...             1  CCR7   \n",
      "negative regulation of dendritic cell apoptotic...             1  CCR7   \n",
      "immunological synapse formation (GO:0001771)                   1  CCR7   \n",
      "\n",
      "                                                   ledge_genes  \n",
      "Term                                                            \n",
      "regulation of interleukin-1 secretion (GO:0050704)        CCR7  \n",
      "positive regulation of ERK1 and ERK2 cascade (G...        CCR7  \n",
      "regulation of pseudopodium assembly (GO:0031272)          CCR7  \n",
      "positive regulation of plasma membrane bounded ...        CCR7  \n",
      "regulation of plasma membrane bounded cell proj...        CCR7  \n",
      "calcium ion transmembrane import into cytosol (...        CCR7  \n",
      "regulation of phosphatidylinositol 3-kinase act...        CCR7  \n",
      "negative regulation of sequestering of calcium ...        CCR7  \n",
      "negative regulation of dendritic cell apoptotic...        CCR7  \n",
      "immunological synapse formation (GO:0001771)              CCR7  \n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gsea = ParetoGSEA(fileObj = 'CD4.csv',nGenes = 25)\n",
    "gsea.topGenesOnly()\n",
    "gsea.runGSEA()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea = ParetoGSEA(fileObj = 'CD4.csv',nGenes = 25)\n",
    "gsea.topGenesOnly()\n",
    "gsea.saveGeneRanks()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I observe the overall network consisting of all genes expressed among CD4 T cells for all archetypes. PDK1, LDHA, MYC, and LEF1, and TCF7 form a linear connection. Based on the previous analysis of the role of MYC and LEF1, PDK1 has a role in cell proliferation and reducing apoptosis in response to hypoxia and oxidative stress, and also functions within pathways involving insulin-like growth factor and insulin signaling. PDK1 has been researched to induce epithilial-mesenchymal transition in various cancer metastasis(4,5) and overall cell invasion and migration. \n",
    "\n",
    "The role of integrins in this networks including receptors ITGA6, ITGB1, and ITGB2 along with the family of chemokines including CCR7, CCR4, CCR5, CCL5,and CXCR6 correspond to integrin activation by chemokines for T-cell adhesion to endothelial cells creating an outside-in signaling from the extra-cellular matrix to the cytoplasm(6). Causes for this mechanism include  tumor invasion and atherosclerosis. Based on this chemokine activation of T cells on the endothelial surface, it can be implicated that this process is within many different organs, thus contributing to viral spread and perhaps blood clots/viral invasions in different organs. Furthermore, it is plausible for matrix degradation to contribute to the virus spread given T cell migration/adhesion within endothelial cells.\n",
    "\n",
    "The network involving IFI27, SAMHD1,ISG15,OAS2, and other genes correspond to anti-viral responses.\n",
    "\n",
    "\n",
    "4. https://www.sciencedirect.com/science/article/abs/pii/S0014482719306299\n",
    "5. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4745712/\n",
    "6. https://pubmed.ncbi.nlm.nih.gov/11005242/\n",
    "\n",
    "\n",
    "<img src=\"CD4.png\" width=\"800\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An archetype analysis gives appropriate biological processes among genes found closest to each arch. The biological processes give insight on Sars-Cov-2 role in the immune system that deviate from the expected innate,adaptive immune responses. With this approach, we can determine which cell type is involved in processes describing virus invasion and spread, and thus can gain a dynamic understanding of key processes occuring within this cell type. Further analysis of interest would be pseudo-time analysis within CD4 T cells to observe cell differentiation within a trajectory. \n",
    "\n",
    "Key Takeaways:\n",
    "\n",
    "1. Integrin family proteins are involved in both chemokine mediated T cell migration and platelet aggregation.\n",
    "\n",
    "2. Proteins including MYC,LEF1,PDK1, and BIRC3 detail mechanisms behind virus spread,virus role in cell protrusion, and mechanisms of epithilial to mesenchymal transition.\n",
    "\n",
    "3. A small network consisting of proteins SYNE1, SYNE2 and CX3CR1 may contribute to dendritic cell chemotaxis and cerebral cortex migration implicating neuronal pathogenesis.\n",
    "\n",
    "4. Certain networks suggest a pro-inflammatory respone induced by MAPK and ERK 1/2 signaling that may lead to a cytokine storm, a known effect of Sars-Cov2 from early findings."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
